sound/soc/sof/intel/hda-loader.c

```diff
@@ -306,49 +312,39 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	if (i == HDA_FW_BOOT_ATTEMPTS) {
 		dev_err(sdev->dev, "error: dsp init failed after %d attempts with err: %d\n",
 			i, ret);
-
-		hda_dsp_dump(sdev, SOF_DBG_REGS | SOF_DBG_PCI | SOF_DBG_MBOX);
-
-		/* disable DSP */
-		snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR,
-					SOF_HDA_REG_PP_PPCTL,
-					SOF_HDA_PPCTL_GPROCEN, 0);
-		goto done;
+		goto cleanup;
 	}
 
 	/* at this point DSP ROM has been initialized and
 	 * should be ready for code loading and firmware boot
 	 */
-	ret = cl_copy_fw(sdev, tag);
-	if (ret < 0)
-		dev_err(sdev->dev, "error: load fw failed ret: %d\n", ret);
-	else
+	ret = cl_copy_fw(sdev, stream);
+	if (!ret)
 		dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
+	else
+		dev_err(sdev->dev, "error: load fw failed ret: %d\n", ret);
 
-done:
-	/* get stream with tag for code loader stream cleanup */
-	list_for_each_entry(s, &bus->stream_list, list) {
-		if (s->direction == SNDRV_PCM_STREAM_PLAYBACK &&
-		    s->stream_tag == tag) {
-			stream = stream_to_hdac_ext_stream(s);
-			break;
-		}
-	}
-
-	if (!stream) {
-		dev_err(sdev->dev,
-			"error: could not get stream with stream tag%d\n",
-			tag);
-		return -ENODEV;
-	}
-
-	/* cleanup code loader stream */
+cleanup:
+	/*
+	 * Perform codeloader stream cleanup.
+	 * This should be done even if firmware loading fails.
+	 */
 	ret1 = cl_cleanup(sdev, &sdev->dmab, stream);
 	if (ret1 < 0) {
 		dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");
-		return ret1;
+		ret = ret1;
 	}
 
+	if (!ret)
+		return ret;
```

Comment:
is there a reason why you want to return here without doing the DSP disable? Or is it a leftover from the previous code before you added the DSP disable?

Reply:
@plbossart we should disable the DSP only in case of errors. if ret is 0, it means fw boot was successful.

Reply:
@ranj063 the diff format is misleading, see the entire code

````c
ret1 = cl_cleanup(sdev, &sdev->dmab, stream);
	if (ret1 < 0) {		if (ret1 < 0) {
		dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");			dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");
		return ret1;			ret = ret1;
	}		}


 	if (!ret)
		return ret;
````

Why would you return if ret1 was not zero? the cleanup failed at this point

Reply:
@plbossart if cleanup fails, ret1 will be negative and I set ret to ret1 and the condition (!ret) will not be true. So we will not return but proceed with disabling the DSP.

Reply:
@plbossart maybe I should add some comments to make it clearer

Reply:
@plbossart I've added some comments to make the error flow a bit clearer. Please let me know if thats helpful.